packages/transaction-controller/src/TransactionController.ts

```diff
@@ -1426,6 +1474,85 @@ export class TransactionController extends BaseController<
 
     return dappSuggestedGasFees;
   }
+
+  /**
+   * Validates and adds external provided transaction to state.
+   *
+   * @param transactionMeta - Nominated external transaction to be added to state.
+   */
+  private async addExternalTransaction(transactionMeta: TransactionMeta) {
+    const { networkId, chainId } = this.getChainAndNetworkId();
+    const fromAddress = transactionMeta?.transaction?.from;
+    const confirmedTxs = this.state.transactions.filter(
+      (transaction) =>
+        transaction.transaction.from === fromAddress &&
+        transaction.status === TransactionStatus.confirmed &&
+        transactionMatchesNetwork(transaction, chainId, networkId),
+    );
+    const pendingTxs = this.state.transactions.filter(
+      (transaction) =>
+        transaction.transaction.from === fromAddress &&
+        transaction.status === TransactionStatus.submitted &&
+        transactionMatchesNetwork(transaction, chainId, networkId),
+    );
+
+    validateConfirmedExternalTransaction(
+      transactionMeta,
+      confirmedTxs,
+      pendingTxs,
+    );
+
+    const updatedTransactions = [...this.state.transactions, transactionMeta];
+    this.update({
+      transactions: this.trimTransactionsForState(updatedTransactions),
+    });
+  }
+
+  /**
+   * Sets other txMeta statuses to dropped if the txMeta that has been confirmed has other transactions
+   * in the transactions have the same nonce.
+   *
+   * @param transactionId - Used to identify original transaction.
+   */
+  private markNonceDuplicatesDropped(transactionId: string) {
+    const { networkId, chainId } = this.getChainAndNetworkId();
+    const transactionMeta = this.getTransaction(transactionId);
+    const nonce = transactionMeta?.transaction?.nonce;
+    const from = transactionMeta?.transaction?.from;
+    const sameNonceTxs = this.state.transactions.filter(
+      (transaction) =>
+        transaction.transaction.from === from &&
+        transaction.transaction.nonce === nonce &&
+        transactionMatchesNetwork(transaction, chainId, networkId),
+    );
+
+    if (!sameNonceTxs.length) {
+      return;
+    }
+
+    // Mark all same nonce transactions as dropped and give it a replacedBy hash
+    sameNonceTxs.forEach((sameNonceTxMeta) => {
```

Comment:
As we're not using the index, or chaining functional logic here, is it simpler to use a standard `for of`?

Reply:
Updated in https://github.com/MetaMask/core/pull/1625/commits/a61549267cb5151f52b1ccca0a2ba78ead7a9f47